File: exercise/python/05_gym_equipment.py
"""
Exercise 5: Gym Equipment Manager

So far you've used basic data types, lists, dictionaries, and functions. Now let's
learn about CLASSES - a way to create your own custom data types! A class is like
a blueprint for creating objects. Think of it as a template for gym equipment.

Concepts covered:
- Classes (blueprints for creating objects)
- Objects (instances of a class)
- Attributes (data stored in an object)
- Methods (functions that belong to a class)
- The __init__ method (constructor)
- The self parameter

Learning Goals:
- Understand the concept of classes and objects
- Create your own class to represent gym equipment
- Use methods to perform actions on objects
- Manage state (data that changes over time)
- Build a gym equipment tracking system

Run this exercise: uv run exercise/python/05_gym_equipment.py
"""

# ============================================================================
# QUICK INTRO: What are classes?
# ============================================================================
# A class is a blueprint for creating objects. An object is a specific instance
# of that blueprint.
#
# Think about it: "Barbell" is a type of equipment (the class). But the specific
# 45-pound Olympic barbell in your gym that's currently being used is an object
# (an instance of the Barbell class).
#
# Classes bundle data (attributes) and functions (methods) together.


# ============================================================================
# EXAMPLE CODE
# ============================================================================

class GymEquipment:
    """
    A class representing a piece of gym equipment.

    Attributes:
        name: The equipment name
        weight: The weight in pounds
        in_use: Whether it's currently being used
        uses_today: How many times it's been used today
    """

    def __init__(self, name, weight):
        """
        The __init__ method is called when you create a new object.
        'self' refers to the specific object being created.

        Parameters:
            name: Equipment name
            weight: Equipment weight in pounds
        """
        self.name = name
        self.weight = weight
        self.in_use = False
        self.uses_today = 0

    def check_out(self):
        """
        Mark equipment as in use.
        Methods are functions that belong to the class.
        They can access and modify the object's attributes using 'self'.
        """
        if self.in_use:
            print(f"{self.name} is already in use!")
        else:
            self.in_use = True
            self.uses_today += 1
            print(f"{self.name} checked out. Uses today: {self.uses_today}")

    def check_in(self):
        """Mark equipment as available."""
        if not self.in_use:
            print(f"{self.name} is already checked in!")
        else:
            self.in_use = False
            print(f"{self.name} checked back in. Ready for next person!")

    def get_status(self):
        """Return a status message about this equipment."""
        status = "IN USE" if self.in_use else "AVAILABLE"
        return f"{self.name} ({self.weight} lbs) - {status} - Used {self.uses_today} times today"


# Create objects (instances) of the GymEquipment class
print("=== Creating Equipment ===")
barbell = GymEquipment("Olympic Barbell", 45)
dumbbell = GymEquipment("50lb Dumbbell", 50)

print(barbell.name)

# Use methods on the objects
print("\n=== Using Equipment ===")
print(barbell.get_status())
barbell.check_out()
print(barbell.get_status())

print()
print(dumbbell.get_status())
dumbbell.check_out()
dumbbell.check_in()
print(dumbbell.get_status())

# Each object has its own separate data
print("\n=== Multiple Uses ===")
barbell.check_in()
barbell.check_out()
barbell.check_in()
barbell.check_out()
print(f"Barbell uses: {barbell.uses_today}")
print(f"Dumbbell uses: {dumbbell.uses_today}")  # Different from barbell!


# ============================================================================
# YOUR TURN: TODO EXERCISES
# ============================================================================

print("\n\n=== YOUR GYM EQUIPMENT MANAGER ===")

# TODO 1: Create objects for your gym
# Create three equipment objects:
# - bench: A "Bench Press Bench" weighing 50 lbs
# - squat_rack: A "Squat Rack" weighing 100 lbs
# - kettlebell: A "32kg Kettlebell" weighing 70 lbs
#
# Then print the status of each using the get_status() method
#
# HINT: Use the same pattern as the examples above
#
# Stuck? Ask: "How do I create an object from a class in Python?"
# Or: "What's the difference between a class and an object?"

# Write your code here:



# TODO 2: Simulate checking out and using equipment
# Simulate someone using the bench:
# 1. Check out the bench
# 2. Print its status
# 3. Check it back in
# 4. Print its status again
#
# Need help? Ask: "How do I call methods on an object?"

# Write your code here:





# TODO 3: Track multiple uses
# Simulate the squat_rack being used 3 times in a row:
# - Check it out, check it in, check it out, check it in, check it out, check it in
# Then print how many times it's been used today (access the uses_today attribute)
#
# HINT: You can access attributes directly: squat_rack.uses_today
#
# Stuck? Ask: "How do I access an object's attributes?"

# Write your code here:





# TODO 4: Create a new class for a Treadmill
# Create a class called Treadmill that has:
# - __init__ method taking: name, max_speed (in mph)
# - Attributes: name, max_speed, is_running (starts as False), distance_today (starts as 0)
# - Method start_running(): sets is_running to True, prints a message
# - Method stop_running(): sets is_running to False, prints a message
# - Method log_distance(miles): adds miles to distance_today
# - Method get_status(): returns a string with treadmill info
#
# HINT: Follow the same pattern as GymEquipment class above
#
# Stuck? Ask: "How do I create a class in Python?"
# Or: "Show me the structure of a class with methods"

# Write your code here:


# TODO 5: Test your Treadmill class
# Create a Treadmill object called treadmill1:
# - Name: "Treadmill #1"
# - Max speed: 12.0 mph
#
# Then:
# 1. Print its status
# 2. Start it running
# 3. Log 3.5 miles
# 4. Stop it running
# 5. Print its status again
#
# Need help? Ask: "How do I use a class I just created?"

# Write your code here:






# ============================================================================
# BONUS CHALLENGES (Optional)
# ============================================================================

# BONUS 1: Add a maintenance tracking feature
# Add a method to the GymEquipment class called needs_maintenance() that returns
# True if uses_today is greater than 10, False otherwise.
# Then add a method called perform_maintenance() that resets uses_today to 0
# and prints a maintenance message.
#
# HINT: You'll need to modify the GymEquipment class definition above
#
# For help: Ask "How do I add new methods to an existing class?"

# Write your code here:




# BONUS 2: Create a BenchPress class that inherits from GymEquipment
# Use class inheritance to create a BenchPress class that:
# - Inherits from GymEquipment
# - Adds a max_weight attribute
# - Adds a method log_lift(weight) that checks if weight > max_weight
#   and updates max_weight if it is
#
# This is advanced! Ask: "How does class inheritance work in Python?"

# Write your code here:




# BONUS 3: Create a gym inventory system
# Create a list called gym_inventory and add all your equipment objects to it.
# Then loop through and print the status of all equipment.
# Calculate and print how many pieces are currently in use.
#
# Want guidance? Ask: "How do I store objects in a list and loop through them?"

# Write your code here:




# ============================================================================
# TESTING YOUR CODE
# ============================================================================

# Expected behavior:
# - bench, squat_rack, and kettlebell should be created and have status messages
# - Equipment should track uses correctly when checked out and in
# - Treadmill class should work with start/stop and distance tracking
# - Each object maintains its own separate state
#
# To test:
# - Print object attributes: print(bench.uses_today)
# - Call methods and verify the output
# - Create multiple objects and verify they're independent
#
# If something's not working: Ask me "Why isn't my [class/method] working?"


if __name__ == "__main__":
    pass

File: exercise/typescript/05_gym_equipment.ts
/**
 * Exercise 5: Gym Equipment Manager with Classes
 *
 * You've mastered functions and interfaces. Now let's learn about CLASSES with TypeScript!
 * Classes combine data (properties) and behavior (methods) into reusable blueprints.
 * TypeScript adds type safety to properties, methods, and access modifiers.
 *
 * Concepts covered:
 * - Class properties with types
 * - Constructor parameters with types
 * - Method parameter and return types
 * - Access modifiers (public, private, readonly)
 * - Creating and using class instances
 * - Type safety in object-oriented code
 *
 * Run this exercise: node exercise/typescript/05_gym_equipment.ts
 */

// ============================================================================
// QUICK INTRO: Classes in TypeScript
// ============================================================================
// A class is a blueprint for creating objects.
//
// class Equipment {
//   name: string;              // Property type
//   weight: number;
//
//   constructor(name: string, weight: number) {  // Constructor types
//     this.name = name;
//     this.weight = weight;
//   }
//
//   display(): void {          // Method return type
//     console.log(this.name);
//   }
// }
//
// const barbell = new Equipment("Barbell", 45);  // TypeScript knows the type!

// ============================================================================
// EXAMPLE CODE
// ============================================================================

/**
 * Represents a piece of gym equipment.
 */
class GymEquipment {
  // Property declarations with types
  name: string;
  weight: number;
  inUse: boolean;
  usesToday: number;

  /**
   * Constructor with typed parameters.
   */
  constructor(name: string, weight: number) {
    this.name = name;
    this.weight = weight;
    this.inUse = false;
    this.usesToday = 0;
  }

  /**
   * Mark equipment as in use.
   * Method with no return value (void).
   */
  checkOut(): void {
    if (this.inUse) {
      console.log(`${this.name} is already in use!`);
    } else {
      this.inUse = true;
      this.usesToday++;
      console.log(`${this.name} checked out. Uses today: ${this.usesToday}`);
    }
  }

  /**
   * Mark equipment as available.
   */
  checkIn(): void {
    if (!this.inUse) {
      console.log(`${this.name} is already checked in!`);
    } else {
      this.inUse = false;
      console.log(`${this.name} checked back in. Ready for next person!`);
    }
  }

  /**
   * Get status message.
   * Method with string return type.
   */
  getStatus(): string {
    const status = this.inUse ? "IN USE" : "AVAILABLE";
    return `${this.name} (${this.weight} lbs) - ${status} - Used ${this.usesToday} times today`;
  }
}

// TypeScript's access modifiers (more advanced)
// Export to make available for exercises
export class SecureEquipment {
  // public: accessible everywhere (default)
  public name: string;

  // private: only accessible inside the class
  private maintenanceCount: number;

  // readonly: can only be set in constructor
  readonly serialNumber: string;

  constructor(name: string, serialNumber: string) {
    this.name = name;
    this.serialNumber = serialNumber;
    this.maintenanceCount = 0;
  }

  // Public method
  public performMaintenance(): void {
    this.maintenanceCount++;
    console.log(`Maintenance performed. Total: ${this.maintenanceCount}`);
  }

  // Private method to reset maintenance count (called when needed)
  resetForNewYear(): void {
    this.maintenanceCount = 0;
    console.log(`Maintenance count reset for ${this.name}`);
  }

  // Getter (like a property but calculated)
  get needsMaintenance(): boolean {
    return this.maintenanceCount < 5;
  }
}

// Create instances
console.log("=== Creating Equipment ===");
const barbell = new GymEquipment("Olympic Barbell", 45);
const dumbbell = new GymEquipment("50lb Dumbbell", 50);

// TypeScript knows the type of barbell!
// Hover over barbell to see: const barbell: GymEquipment

console.log("\n=== Using Equipment ===");
console.log(barbell.getStatus());
barbell.checkOut();
console.log(barbell.getStatus());

console.log();
dumbbell.checkOut();
dumbbell.checkIn();
console.log(dumbbell.getStatus());

// Each instance has its own data
console.log("\n=== Multiple Uses ===");
barbell.checkIn();
barbell.checkOut();
barbell.checkIn();
barbell.checkOut();
console.log(`Barbell uses: ${barbell.usesToday}`);
console.log(`Dumbbell uses: ${dumbbell.usesToday}`);

// ============================================================================
// YOUR TURN: TODO EXERCISES
// ============================================================================

console.log("\n\n=== YOUR GYM EQUIPMENT MANAGER ===");

// TODO 1: Create three equipment objects
// Create instances with proper types:
// - bench: type GymEquipment, "Bench Press Bench", 50 lbs
// - squatRack: type GymEquipment, "Squat Rack", 100 lbs
// - kettlebell: type GymEquipment, "32kg Kettlebell", 70 lbs
//
// Type hint: const varName: ClassName = new ClassName(args);
// OR let TypeScript infer: const varName = new ClassName(args);
//
// Stuck? Ask: "How do I create an instance of a class in TypeScript?"

// Write your code here:

// TODO 2: Use the getStatus method
// Call getStatus() on each of your three equipment objects and print the results
//
// Type hint: objectName.methodName()
//
// Need help? Ask: "How do I call methods on a class instance?"

// Write your code here:

// TODO 3: Simulate using equipment
// For the bench:
// 1. Check it out
// 2. Print its status
// 3. Check it back in
// 4. Print its status again
//
// Type hint: Call the checkOut(), getStatus(), and checkIn() methods
//
// Stuck? Ask: "How do I call multiple methods on an object in sequence?"

// Write your code here:

// TODO 4: Track multiple uses
// Check the squatRack out and in 3 times (simulate 3 people using it)
// Then print how many times it's been used today
//
// HINT: Access the usesToday property directly: squatRack.usesToday
//
// Type hint: TypeScript knows usesToday is a number!
//
// Need help? Ask: "How do I access a class property in TypeScript?"

// Write your code here:

// TODO 5: Create a Treadmill class
// Create a class called Treadmill with:
// - Properties (with types!):
//   - name: string
//   - maxSpeed: number
//   - isRunning: boolean (initialize to false)
//   - distanceToday: number (initialize to 0)
//
// - Constructor:
//   - Parameters: name (string), maxSpeed (number)
//   - Initialize all properties
//
// - Methods:
//   - startRunning(): void - sets isRunning to true, prints message
//   - stopRunning(): void - sets isRunning to false, prints message
//   - logDistance(miles: number): void - adds miles to distanceToday
//   - getStatus(): string - returns status with name, speed, running state, distance
//
// Type hint: Follow the GymEquipment class pattern above
//
// Stuck? Ask: "How do I create a class with typed properties and methods in TypeScript?"

// Write your code here:

// TODO 6: Create and test a Treadmill instance
// Create a Treadmill object:
// - Name: "Treadmill #1"
// - Max speed: 12.0
//
// Then:
// 1. Print its status
// 2. Start it running
// 3. Log 3.5 miles
// 4. Stop it running
// 5. Print status again
//
// Type hint: const treadmill = new Treadmill("Treadmill #1", 12.0);
//
// Need help? Ask: "How do I create and use a custom class?"

// Write your code here:

// ============================================================================
// BONUS CHALLENGES (Optional)
// ============================================================================

// BONUS 1: Add access modifiers to Treadmill
// Modify your Treadmill class to use:
// - public for name and getStatus()
// - private for distanceToday
// - readonly for maxSpeed (set only in constructor)
//
// Add a getter for distance: get distance(): number { return this.distanceToday; }
// This lets you read the private property safely!
//
// For help: Ask "How do access modifiers work in TypeScript classes?"

// Write your code here:

// BONUS 2: Create a class that implements an interface
// First, create an interface:
// interface Equipment {
//   name: string;
//   getStatus(): string;
// }
//
// Then create a BenchPress class that implements Equipment:
// class BenchPress implements Equipment {
//   name: string;
//   maxWeight: number;
//   // ... implement required interface members plus your own
// }
//
// Want guidance? Ask: "How do I make a class implement an interface in TypeScript?"

// Write your code here:

// BONUS 3: Create a static method
// Add a static method to your Treadmill class:
// static convertMilesToKm(miles: number): number {
//   return miles * 1.60934;
// }
//
// Static methods are called on the class itself, not instances:
// const km = Treadmill.convertMilesToKm(5);
//
// Need help? Ask: "What are static methods in TypeScript?"

// Write your code here:

// ============================================================================
// TESTING YOUR CODE
// ============================================================================

// Expected behavior:
// - Three equipment objects created (bench, squatRack, kettlebell)
// - Status shows equipment details and usage count
// - Check-out/check-in changes inUse status
// - usesToday increments with each check-out
// - Treadmill class works with all methods
// - Each object maintains its own state
//
// TypeScript provides:
// - Type checking for property access
// - Auto-completion for methods and properties
// - Errors when accessing non-existent members
// - Type safety for method parameters and returns
//
// Key learning points:
// - Properties need type annotations
// - Constructor parameters need types
// - Methods need parameter and return types
// - Access modifiers control visibility (public, private, readonly)
// - TypeScript ensures type safety for classes
//
// If you get type errors:
// - Check all properties have type annotations
// - Verify method parameters have types
// - Ensure return types match what you're returning
// - Check that you're accessing public properties/methods
//
// Ask me: "Why is TypeScript showing an error in my class?"

// Make this file a module to avoid variable name conflicts with other exercise files
export {};
